firedrake/bcs.py

```diff
@@ -450,10 +450,15 @@ def apply(self, r, u=None):
             r = r.sub(idx)
             if u:
                 u = u.sub(idx)
+
+        bc_checkpointed = self.block_variable.checkpoint
+        bc = bc_checkpointed.checkpoint \
+            if bc_checkpointed is not None else self.function_arg
+
         if u:
-            r.assign(u - self.function_arg, subset=self.node_set)
+            r.assign(u - bc, subset=self.node_set)
```

Comment:
It is still missing a test for this case.

Reply:
You will get this case by passing `pre_apply_bcs=False` to the NLVS

Reply:
Thank you.